Support Spring's new RestClient with auto configuration (#3198)

nandorholozsnyak opened the pull request:
- Spring RestClient support

## :scroll: Description
- New classes are added to auto configure the neccessary interceptors for newly created RestClients
- Tests are added to test the new functionality.


## :bulb: Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- The RestClient is similar to the RestTemplate but uses different ways of creation and because of that the older auto configurations did not put the required interceptors into these new RestClient instances

## :green_heart: How did you test it?
- With the unit and IT tests

## :pencil: Checklist
<!--- Put an `x` in the boxes that apply -->

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [ ] No breaking change or entry added to the changelog.
- [ ] No breaking change for hybrid SDKs or communicated to hybrid SDKs.


## :crystal_ball: Next steps

nandorholozsnyak commented on sentry-spring-boot-jakarta/src/test/kotlin/io/sentry/spring/boot/jakarta/SentrySpanRestClientCustomizerTest.kt:
```diff
@@ -0,0 +1,288 @@
+    }
+
+    @Test
+    fun `when transaction is active and throws IO exception, creates span with error status around RestClient HTTP call`() {
```
This test hangs locally, so I will fix it soon.

nandorholozsnyak commented on sentry-spring-boot-jakarta/src/main/java/io/sentry/spring/boot/jakarta/SentrySpanRestClientCustomizer.java:
```diff
@@ -0,0 +1,28 @@
+
+  @Override
+  public void customize(final @NotNull RestClient.Builder restClientBuilder) {
+    restClientBuilder.requestInterceptors(clientHttpRequestInterceptors -> {
```
This is a tricky one, I saw the same concept in the `SentrySpanRestTemplateCustomizer` but I'm not sure if the contains can catch that a `SentrySpanClientHttpRequestInterceptor` is already registered or not, as it is being instantiated in the constructor, so different instances can be registered easily. We might need a proper equals in the `SentrySpanClientHttpRequestInterceptor`

nandorholozsnyak commented:
> Thanks for the PR @nandorholozsnyak, we'll take a look but it might take a while as I'm feeling a bit under the weather.

Hello there,

Sorry for the late reply, but I could only solve the failing test today. Now I would like to ask for a review from you guys. I'll try to sync with master.

adinauer reviewed (approved):
Thanks again for the PR and also your patience. I've added a changelog, added RestClient to our sample and tweaked the reported trace origin.

I've opened a PR for docs: https://github.com/getsentry/sentry-docs/pull/9531 which we should merge after this PR has been released.

nandorholozsnyak commented:
> Thanks again for the PR and also your patience. I've added a changelog, added RestClient to our sample and tweaked the reported trace origin.
> 
> I've opened a PR for docs: [getsentry/sentry-docs#9531](https://github.com/getsentry/sentry-docs/pull/9531) which we should merge after this PR has been released.

Thank you so much.
